Source/Diagnostics/ReducedDiags/Costs.cpp

```diff
@@ -0,0 +1,199 @@
+    AMREX_ALWAYS_ASSERT_WITH_MESSAGE(cost_heuristic != nullptr,
+        "Costs reduced diagnostic does not work with timer-based costs update.");
+
+    // judge if the diags should be done
+    // costs is initialized only if we're doing load balance
+    if ( ((step+1) % m_freq != 0) || warpx.get_load_balance_int() < 1 ) { return; }
+
+    // get number of boxes over all levels
+    auto nLevels = warpx.finestLevel() + 1;
+    int nBoxes = 0;
+    for (int lev = 0; lev < nLevels; ++lev)
+    {
+        const amrex::Vector<amrex::Real>* cost_heuristic = warpx.getCostsHeuristic(lev);
+        nBoxes += cost_heuristic->size();
+    }
+
+    // keep track of the max number of boxes, this is needed later on to fill
+    // the jagged array (in case each step does not have the same number of boxes)
+    nBoxesMax = std::max(nBoxesMax, nBoxes);
+
+    // resize and clear data array
+    m_data.resize(nDataFields*nBoxes, 0.0);
+    m_data.assign(nDataFields*nBoxes, 0.0);
+
+    // if not a load balance step, we must compute costs
+    if ((step+1) % warpx.get_load_balance_int() != 0)
+    {
+        if (warpx.load_balance_costs_update_algo == LoadBalanceCostsUpdateAlgo::Heuristic)
+        {
+            warpx.ComputeCostsHeuristic();
+
+        } else if (WarpX::load_balance_costs_update_algo == LoadBalanceCostsUpdateAlgo::Timers)
+        {
+            // costs update via timers, nothing to compute
+        }
+    }
+
+    // keeps track of correct index in array over all boxes on all levels
+    int shift = 0;
+
+    // save data
+    for (int lev = 0; lev < nLevels; ++lev)
+    {
+        const amrex::DistributionMapping& dm = warpx.DistributionMap(lev);
+        const amrex::Vector<amrex::Real>* cost_heuristic = warpx.getCostsHeuristic(lev);
+        const MultiFab & Ex = warpx.getEfield(lev,0);
+        for (MFIter mfi(Ex, false); mfi.isValid(); ++mfi)
+        {
+            const Box& tbx = mfi.tilebox();
+            m_data[shift + mfi.index()*nDataFields + 0] = (*cost_heuristic)[mfi.index()];
+            m_data[shift + mfi.index()*nDataFields + 1] = dm[mfi.index()];
+            m_data[shift + mfi.index()*nDataFields + 2] = lev;
+            m_data[shift + mfi.index()*nDataFields + 3] = tbx.loVect()[0];
+            m_data[shift + mfi.index()*nDataFields + 4] = tbx.loVect()[1];
+            m_data[shift + mfi.index()*nDataFields + 5] = tbx.loVect()[2];
+        }
+
+        // we looped through all the boxes on level lev, update the shift index
+        shift += nDataFields*(cost_heuristic->size());
+    }
+
+    // parallel reduce to IO proc and get data over all procs
+    amrex::Vector<amrex::Real>::iterator it_m_data = m_data.begin();
+    amrex::Real* addr_it_m_data = &(*it_m_data);
+    ParallelReduce::Sum(addr_it_m_data,
+                        m_data.size(),
+                        ParallelDescriptor::IOProcessorNumber(),
+                        ParallelContext::CommunicatorSub());
```

Comment:
Can't this give the same?
```
ParallelDescriptor::ReduceRealSum(m_data.data(), m_data.size(), ParallelDescriptor::IOProcessorNumber());
```

Reply:
Yes, that also works!  I switched to that.